### DISCUSSION TITLE
Why does JwtCirce.decode lead to java.time.Clock validating the expiration?

### DISCUSSION THREAD

**Discussion by drewboardman** [jwt-scala/jwt-scala#504] [General]
The `JwtCirce.decode` function does not implicitly take an instance of `java.time.Clock`

https://github.com/jwt-scala/jwt-scala/blob/c73b5715ac2a812e870c54e7cace016d0e7f01ff/core/src/main/scala/JwtCore.scala#L809

```Scala
  def decode(token: String, key: String, algorithms: Seq[JwtHmacAlgorithm]): Try[C] =
    decode(token, key, algorithms, JwtOptions.DEFAULT)
```

However, this leads to the function `JwtTime.validateNowIsBetween`

https://github.com/jwt-scala/jwt-scala/blob/c73b5715ac2a812e870c54e7cace016d0e7f01ff/core/src/main/scala/JwtTime.scala#L60

```Scala
  def validateNowIsBetween(start: Option[Long], end: Option[Long])(implicit
      clock: Clock
  ): Try[Unit] = {
    val timeNow = now

    (start, end) match {
      case (Some(s), _) if s > timeNow  => Failure(new JwtNotBeforeException(s))
      case (_, Some(e)) if e <= timeNow => Failure(new JwtExpirationException(e))
      case _                            => Success(())
    }
  }
  ```

This is unexpected behavior. I've tried searching through the flow of this decoding, and I can't find where this `Clock` instance gets slipped in.

Is this intentional? The documentation seems to infer that `JwtCirce` simply decodes the JWT, and makes no assumptions about its expiry.

**Comment by drewboardman** [comment]
It looks like you're inheriting it from `JwtCore`

```Scala
trait JwtCore[H, C] {
  implicit private[jwt] def clock: Clock
  ...
}
trait JwtJsonCommon[J, H, C] extends JwtCore[H, C] 
...
trait JwtCirceParser[H, C] extends JwtJsonCommon[Json, H, C]
```

The wording of `decode` is confusing in this instance. If the `java.time.Clock` is a dependency of `JwtCirce.decode`, the client should be able to use their instance - instead of this embedded instance that can't be altered.

**Comment by Abhijeetmishr** [comment]
@drewboardman if you can guide can I pick up this issue ?

**Comment by erwan** [comment]
The fact that `JwtCirce.decode` does not take an implicit Clock is not really an issue in itself, as Drew explained.

I'm moving it to discussions so you can discuss the problem you're facing.
